### PR TITLE
[Refactor] Convert hex string to Data

### PR DESCRIPTION
**Pull request by KaterinaWire**
## What's new in this PR?

1. Added a method for converting a hexadecimal string to Data.
2. Added Data initializer.
3. Tests.

### CODE REVIEW

File: Source/Public/Data+ZMSCrypto.swift [hunk 1]
```diff
@@ -22,6 +22,15 @@ import Foundation
 // Mapping of @c NSData helper methods to Swift 3 @c Data. See original methods for description.
 public extension Data {
     
+    private static let hexRegex = try! NSRegularExpression(pattern: "^([a-fA-F0-9][a-fA-F0-9])*$", options: [])
+
+    init?(hexString: String) {
+        if Data.hexRegex.matches(in: hexString, range: NSMakeRange(0, hexString.count)).isEmpty {
```
Comment (johnxnguyen) [inline Source/Public/Data+ZMSCrypto.swift:28]: Better to use `hexString.utf16.count` when using `NSRange`.

File: Source/Public/String+HexDecodedData.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public extension String {
+    
+    /// A data representation of the hexadecimal bytes in this string.
+    func zmHexDecodedData() -> Data {
```
Comment (johnxnguyen) [inline Source/Public/String+HexDecodedData.swift:24]: So the `Data` initializer is fallible, but this extension method is not. It makes more sense to make this return `Data?` and move the hex string validation here.

File: Tests/Source/NSData+ZMSCryptoTests.swift [hunk 3]
```diff
@@ -296,4 +296,42 @@ extension NSData_ZMSCryptoTests {
...
+        
+        // when
+        let decoded = Data(hexString: normalString)
+        
+        // then
+        let array : Array<UInt8> = Array(0...255)
+        let expectedData = Data(array)
+        
+        XCTAssertNil(decoded)
+        XCTAssertNotEqual(decoded, expectedData)
```
Comment (johnxnguyen) [inline Tests/Source/NSData+ZMSCryptoTests.swift:334]: I think asserting for `nil` is sufficient.

**Review by johnxnguyen — CHANGES_REQUESTED** [review]
Your hex string validator allows empty strings. Does this mean the data return is also empty? I would add some more tests to check some edge cases, including empty strings, hex strings with uppercase characters, hex strings with lowercase characters, hex strings with a mix of upper and lowercase, and hex strings with an uneven number of characters.

File: Source/Public/String+HexDecodedData.swift (hunk 2 above)
Comment (KaterinaWire) [reply Source/Public/String+HexDecodedData.swift]: Done

**Comment by KaterinaWire** [comment]
> Your hex string validator allows empty strings. Does this mean the data return is also empty? I would add some more tests to check some edge cases, including empty strings, hex strings with uppercase characters, hex strings with lowercase characters, hex strings with a mix of upper and lowercase, and hex strings with an uneven number of characters.

@johnxnguyen I added tests

**Review by johnxnguyen — APPROVED** [review]
